orca/call_metric_recorder.go

```diff
@@ -0,0 +1,122 @@
+ *
+ */
+
+package orca
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	v3orcapb "github.com/cncf/xds/go/xds/data/orca/v3"
+)
+
+// CallMetricRecorder provides functionality to record per-RPC custom backend
+// metrics. See CallMetricsServerOption() for more details.
+//
+// Safe for concurrent use.
+type CallMetricRecorder struct {
+	cpu    atomic.Value // float64
+	memory atomic.Value // float64
+
+	mu          sync.RWMutex
+	requestCost map[string]float64
+	utilization map[string]float64
+}
+
+func newCallMetricRecorder() *CallMetricRecorder {
+	return &CallMetricRecorder{
+		requestCost: make(map[string]float64),
+		utilization: make(map[string]float64),
+	}
+}
+
+// SetCPUUtilization records a measurement for the CPU utilization metric.
+func (c *CallMetricRecorder) SetCPUUtilization(val float64) {
+	c.cpu.Store(val)
+}
+
+// SetMemoryUtilization records a measurement for the memory utilization metric.
+func (c *CallMetricRecorder) SetMemoryUtilization(val float64) {
+	c.memory.Store(val)
+}
+
+// SetRequestCost records a measurement for a request cost metric,
+// uniquely identifiable by name.
+func (c *CallMetricRecorder) SetRequestCost(name string, val float64) {
+	c.mu.Lock()
+	c.requestCost[name] = val
+	c.mu.Unlock()
+}
+
+// SetUtilization records a measurement for a utilization metric uniquely
+// identifiable by name.
+func (c *CallMetricRecorder) SetUtilization(name string, val float64) {
+	c.mu.Lock()
+	c.utilization[name] = val
+	c.mu.Unlock()
+}
+
+// toLoadReportProto dumps the recorded measurements as an OrcaLoadReport proto.
+func (c *CallMetricRecorder) toLoadReportProto() *v3orcapb.OrcaLoadReport {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	cost := make(map[string]float64, len(c.requestCost))
+	for k, v := range c.requestCost {
+		cost[k] = v
+	}
+	util := make(map[string]float64, len(c.utilization))
+	for k, v := range c.utilization {
+		util[k] = v
+	}
+	cpu, _ := c.cpu.Load().(float64)
+	mem, _ := c.memory.Load().(float64)
+	return &v3orcapb.OrcaLoadReport{
+		CpuUtilization: cpu,
+		MemUtilization: mem,
+		RequestCost:    cost,
+		Utilization:    util,
+	}
+}
+
+type callMetricRecorderCtxKey struct{}
+
+// CallMetricRecorderFromContext returns the RPC specific custom metrics
+// recorder [CallMetricRecorder] embedded in the provided RPC context.
+//
+// Returns nil if no custom metrics recorder is found in the provided context,
+// which will be the case when custom metrics reporting is not enabled.
+func CallMetricRecorderFromContext(ctx context.Context) *CallMetricRecorder {
+	// The actual value stored in the context is the address of the pointer to a
+	// CallMetricRecorder. This is required for the lazy allocation of the
+	// metric recorder, happening in here, which modifies the pointer stored in
+	// the address value found in the context.
+	addr, ok := ctx.Value(callMetricRecorderCtxKey{}).(**CallMetricRecorder)
+	if !ok {
+		return nil
+	}
+	if *addr == nil {
+		*addr = newCallMetricRecorder()
+	}
```

Comment:
This looks racy; I think we'll need a `sync.Once` in a wrapping struct, or something similar.

Reply:
Done. Thanks.